lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+newSharedState accXPub accIx g pTemplate dTemplate =
+    extendSharedState (toEnum 0) $
+    unsafeSharedState accXPub accIx g pTemplate dTemplate Map.empty
+
+replaceCosignersWithVerKeys
+    :: ScriptTemplate
+    -> Int
+    -> Script KeyHash
+replaceCosignersWithVerKeys (ScriptTemplate xpubs scriptTemplate) index =
+    replaceCosigner scriptTemplate
+  where
+    replaceCosigner :: Script Cosigner -> Script KeyHash
+    replaceCosigner = \case
+        RequireSignatureOf c -> RequireSignatureOf $ toKeyHash c
+        RequireAllOf xs      -> RequireAllOf (map replaceCosigner xs)
+        RequireAnyOf xs      -> RequireAnyOf (map replaceCosigner xs)
+        RequireSomeOf m xs   -> RequireSomeOf m (map replaceCosigner xs)
+        ActiveFromSlot s     -> ActiveFromSlot s
+        ActiveUntilSlot s    -> ActiveUntilSlot s
+    toKeyHash :: Cosigner -> KeyHash
+    toKeyHash c =
+        let ix = toEnum index
+            (Just accXPub) = liftXPub <$> Map.lookup c xpubs
+            verKey = deriveMultisigPublicKey accXPub ix
+        in hashKey verKey
+
+constructAddressFromIx
+    :: CA.NetworkTag
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Index 'Soft 'ScriptK
+    -> Address
+constructAddressFromIx tag pTemplate dTemplate ix =
+    let delegationCredential = DelegationFromScript . toScriptHash
+        paymentCredential = PaymentFromScript . toScriptHash
+        createAddress pScript' dScript' =
+            CA.unAddress $
+            delegationAddress tag
+            (paymentCredential pScript') (delegationCredential dScript')
+        dTemplate' = fromMaybe pTemplate dTemplate
+        pScript = replaceCosignersWithVerKeys pTemplate (fromEnum ix)
+        dScript = replaceCosignersWithVerKeys dTemplate' (fromEnum ix)
+    in Address $ createAddress pScript dScript
+
+addNotDiscoveredAddressToMap
+    :: CA.NetworkTag
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Index 'Soft 'ScriptK
+    -> Map Address (Index 'Soft 'ScriptK, AddressState)
+    -> Map Address (Index 'Soft 'ScriptK, AddressState)
+addNotDiscoveredAddressToMap tag pTemplate dTemplate ix currentMap =
+    let address = constructAddressFromIx tag pTemplate dTemplate ix
+    in Map.insert address (ix, Unused) currentMap
+
+templatesComplete
+    :: ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Bool
+templatesComplete pTemplate dTemplate =
+    isRight (validateScriptTemplate RequiredValidation pTemplate) &&
+    case dTemplate of
+        Nothing -> True
+        Just dTemplate' ->
+            isRight (validateScriptTemplate RequiredValidation dTemplate')
+
+keyHashFromAccXPubIx
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+keyHashFromAccXPubIx accXPub ix =
+    hashVerificationKey $ deriveVerificationKey accXPub ix
```

Comment:
Careful that keys on the payment template are derived with role=3, but keys on the delegation template are derive with role=4. I'd suggest to make the role explicit here to avoid shooting ourselves in the foot.

Reply:
corrected here https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/9522d0c0a5b0fd39ffab1b03235f364c78a972f3